### PR TITLE
FQDN: Do not lock twice the endpoint Manager.

### PR DESCRIPTION
**Pull request by eloycoto**
In DNS proxy all the information about the endpoint is already gathered
from the endpoint manager in the ServeDNS function, this function called
a few times to NotifyDNSMsg, and this function also lock the endpoint
manager to retrieve endpoint.

This commit send the endpointID and the endpointAddress in the callback,
so the number of locks in the EndpointManager decreases and make the
dns-proxy a bit faster. Also the lookups are using the endpointManager
map instead of the IP, so are a bit faster.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7508)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by ianvernon** [comment]
I might be missing something, but it appears as if the same number of `endpointmanager` lookups are performed here?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 1]
```diff
@@ -118,9 +118,16 @@ type DNSProxy struct {
 // See DNSProxy.LookupEndpointIDByIP for usage.
 type LookupEndpointIDByIPFunc func(ip net.IP) (endpointID string, err error)
 
+// EndpointNotifyContext is a struct that stores the DNS request/response
+// context related to an endpoint.
+type EndpointNotifyContext struct {
+	EpAddr string
```
Comment (aanm) [inline pkg/fqdn/dnsproxy/proxy.go]: Why isn't this a `net.IP`?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 2]
```diff
@@ -118,9 +118,16 @@ type DNSProxy struct {
...
+	EpID   int64
```
Comment (aanm) [inline pkg/fqdn/dnsproxy/proxy.go]: s/int64/uint64

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
test-me-please

**Review by raybejjani — CHANGES_REQUESTED** [review]
@eloycoto Sorry, I wasn't clear in what I was proposing. I was expecting that we could remove LookupEndpointByID from ServeDNS completely. We would keep a pointer to the endpoint inside DNSProxy that we update in `DNSProxy.UpdateAllowed`. The idea was to avoid locking during `ServeDNS`, so this might also mean caching some non-changing data on the endpoint (like its IPs) but `Endpoint.ID` might be enough for all the logs we use.

EDIT: We chatted offline. eloycoto pointed out that keeping a copy of the endpoint in the proxy would likely have more raciness than just doing a single lookup in endpoint manager. We can do that and ditch the EpContext type, though, since the Endpoint.ID field is available on the endpoint objects.
Separately, there seems to be a nil-dereference in some logging code when an endpoint can't be found that we'll also fix.

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go [hunk 3]
```diff
@@ -275,40 +276,38 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	requestID := request.Id // keep the original request ID
 	qname := string(request.Question[0].Name)
 	protocol := w.LocalAddr().Network()
-	endpointAddr := w.RemoteAddr().String()
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.DNSName:      qname,
 		logfields.IPAddr:       w.RemoteAddr(),
 		logfields.DNSRequestID: request.Id})
 
-	scopedLog.Debug("Handling DNS query from endpoint")
```
Comment (raybejjani) [inline pkg/fqdn/dnsproxy/proxy.go]: Could you leave this debug line in? It's useful when diagnosing whether the proxy is being slow to begin handling packets or something else is slow in NotifyOnDNSMsg.

**Review by raybejjani — APPROVED** [review]
Apart from one debug message, LGTM

**Comment by eloycoto** [comment]
test-me-please (etcd-issues https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/11148/console)

**Comment by eloycoto** [comment]
test-me-please

**Pull request merged**